### DISCUSSION TITLE
Understanding the output of the weather forecast api

### DISCUSSION THREAD

**Discussion by JamisonProctor** [open-meteo/open-meteo#399] [Q&A]
Hey everyone,

I'm a beginner data scientist and have been getting pretty frustrated with the accuracy of weather forecasts in my local area. So, I thought it'd be a fun project to analyze the accuracy of these forecasts and maybe come up with a way to communicate them more clearly.

I collected current weather and forecast information for a few months using the open-meteo API and started my analysis. But honestly, I'm a bit confused by the results. From what I understand (please correct me if I'm wrong), the forecast data from each API call provides the weather predictions for the upcoming hours: the first position for one hour from now, the second position for two hours from now, and so on, up to the 100th position for 100 hours from now.

To make sense of the data, I tried adjusting the data frame so that each forecast lines up with others for the same end time. Here's the code snippet I used for that:

`for i in range(df.shape[1]):
      df.iloc[:,i] = df.iloc[:,i].shift(i)`

Well, I had this theory that the further-out forecasts would be less accurate, and as we get closer to the forecasted time, they would improve. But guess what? It looks like my theory didn't quite pan out.

I've been digging into the data from the open-meteo API, analyzing forecasts for different timeframes. Surprisingly, I'm not seeing a clear trend of increasing accuracy as we approach the forecasted time. The forecasts further out aren't necessarily worse, and those closer in time don't always show a significant boost in accuracy.

![output](https://github.com/open-meteo/open-meteo/assets/36643682/d7d03c5c-0229-45c7-973e-f7010b288372)

The data doesn't seem to flatten into a more accurate forecast as I expected, and it appears to vary depending on the time of day the forecast was made. Surprisingly, it looks quite similar to what I see without using the above code to adjust the data frame. 

![output](https://github.com/open-meteo/open-meteo/assets/36643682/8a023f6a-1641-45b8-be3f-6e4e189c4785)

I'm guessing there might be an issue with my code. So, if any of you have any advice, clarification, or even a link to a more detailed explanation, I'd totally appreciate it. Thanks!

**Comment by patrick-zippenfenig** [comment]
Hi there, for which coordinates are you analyzing the forecast? To which data are you comparing jr?

Yes, your assumption should be true that the forecast further out is less accurate.

**Comment by JamisonProctor** [comment]
Hi @patrick-zippenfenig - Thanks for getting back to me, and for the access to the awesome API. 

I am in Munich Germany, so I use the coordinates latitude=48.14 longitude=11.58. Currently I am not comparing any data. I am just not clear on why what I would like to do does not seem to work. 

I made something in excel to try to illustrate my issue. Below is my understanding of the API call result when I call it every hour. 

![Screenshot 2023-07-20 at 19 09 47](https://github.com/open-meteo/open-meteo/assets/36643682/84c50c2d-2652-4dbb-bf93-5abf216fc6f1)

My idea is to use the below pandas code to shift the forecast for each hour so that each forecast is in a row with all of the other forecasts of the same target time. 

`for i in range(df.shape[1]): 
     df.iloc[:,i] = df.iloc[:,i].shift(i)`

I expect the result to be as follows, with each forecast for the same target time in the same row. 

![Screenshot 2023-07-20 at 19 17 12](https://github.com/open-meteo/open-meteo/assets/36643682/c1a18558-f3df-4ecf-9f16-442bd81dfd0c)

However, when I plot the output where I would expect to get a converging line which stabilizes to very close to the actual weathers as the target time becomes closer, instead I get the below image. Please note that this is temperature forecast data.

![image](https://github.com/open-meteo/open-meteo/assets/36643682/d51936e7-0bf4-47f3-97e9-27f19d5434f6)

The forecast seems to move up and down considerably, and also follows a daytime/nighttime temperature pattern. In fact the output without the above column shift adjustment similarly shows a daytime/nighttime temperature variance pattern.

![image](https://github.com/open-meteo/open-meteo/assets/36643682/07e8b257-c087-4627-aef0-56f0a06aebb9)

Is this strange? It seems strange, but I am also a beginner and very likely have made a mistake. 

Is my understanding illustrated in the above excel images correct? If so, does the forecast for a future target somehow depend on the temperature of location for the time that the API is called? (i.e. if I call the API at noon, does it somehow take the current noon time temperature into a forecast in 12 hours?)

I am very sorry to bother you when you obviously have more important things to do, but any help you can provide would be greatly appreciated.

**Reply by patrick-zippenfenig** [reply]
Data always starts at 0:00 UTC (or localtime if a timezone is selected). It looks to me like your shift operation is not working correctly. I would recommend to use the `time` column to ensure you are matching data correctly. 

Please note that I also want to keep data from previous weather model runs to evaluate model performance (see #206). This feature will be implemented in the next months.

**Reply by JamisonProctor** [reply]
@patrick-zippenfenig I've overlooked the data always starting at 0:00 UTC. I will keep grinding at it. Thank you!
